edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs

```diff
@@ -261,28 +261,38 @@ private async Task CleanQueue(bool checkEntireQueueOnCleanup)
                 {
                     foreach (KeyValuePair<string, ISequentialStore<MessageRef>> endpointSequentialStore in this.messageStore.endpointSequentialStores)
                     {
+                        var messageQueueId = endpointSequentialStore.Key;
                         try
                         {
                             if (this.cancellationTokenSource.IsCancellationRequested)
                             {
                                 return;
                             }
 
-                            Events.CleanupTaskStarted(endpointSequentialStore.Key);
-                            CheckpointData checkpointData = await this.messageStore.checkpointStore.GetCheckpointDataAsync(endpointSequentialStore.Key, CancellationToken.None);
+                            var (endpointId, priority) = MessageQueueIdHelper.ParseMessageQueueId(messageQueueId);
+                            Events.CleanupTaskStarted(messageQueueId);
+                            CheckpointData checkpointData = await this.messageStore.checkpointStore.GetCheckpointDataAsync(messageQueueId, CancellationToken.None);
                             ISequentialStore<MessageRef> sequentialStore = endpointSequentialStore.Value;
-                            Events.CleanupCheckpointState(endpointSequentialStore.Key, checkpointData);
+                            Events.CleanupCheckpointState(messageQueueId, checkpointData);
                             int cleanupEntityStoreCount = 0;
 
+                            // If cleanup just peek head, message counts is tailOffset-headOffset+1
+                            // otherwise count while iterating over the queue.
+                            var headOffset = 0L;
+                            var tailOffset = sequentialStore.GetTailOffset(CancellationToken.None);
+                            var messageCounts = 0L;
+
                             async Task<bool> DeleteMessageCallback(long offset, MessageRef messageRef)
                             {
-                                var enqueuedTime = DateTime.UtcNow - messageRef.TimeStamp;
-                                if (checkpointData.Offset < offset &&
-                                     enqueuedTime < messageRef.TimeToLive)
+                                var expiry = messageRef.TimeStamp + messageRef.TimeToLive;
+                                if (offset > checkpointData.Offset && expiry > DateTime.UtcNow)
```

Comment:
When would offset be less than checkpointData.Offset?

Reply:
`CheckPointer` will update  `checkpointData.Offset` which indicates message get sent. But `offset` here is message ref from database which won't be update unless deleted. They are completely separate.